### PR TITLE
Smh enterprise refactor

### PR DESCRIPTION
**Pull request by ilackarms**
refactor code including the extensions api to enable the enterprise extensions server
BOT NOTES: 
resolves https://github.com/solo-io/service-mesh-hub/issues/1049

### CODE REVIEW

File: pkg/mesh-networking/translation/istio/extensions/istio_networking_extender.go [hunk 1]
```diff
@@ -3,28 +3,23 @@ package extensions
...
-	// PatchTrafficTargetOutputs retrieves from the NetworkingExtensionsServers and applies patches to the outputs for a given TrafficTarget
-	PatchTrafficTargetOutputs(ctx context.Context, trafficTarget *v1alpha2.TrafficTarget, trafficTargetOutputs istio.Builder) error
-
-	// PatchWorkloadOutputs retrieves from the NetworkingExtensionsServers and applies patches to the outputs for a given Workload
-	PatchWorkloadOutputs(ctx context.Context, workload *v1alpha2.Workload, workloadOutputs istio.Builder) error
-
-	// PatchMeshOutputs retrieves from the NetworkingExtensionsServers and applies patches to the outputs for a given Mesh
-	PatchMeshOutputs(ctx context.Context, mesh *v1alpha2.Mesh, meshOutputs istio.Builder) error
+	// PatchOutputs retrieves from the NetworkingExtensionsServers and applies patches to the outputs for a given TrafficTarget
+	PatchOutputs(ctx context.Context, inputs input.Snapshot, outputs istio.Builder) error
```
Comment (ilackarms) [inline pkg/mesh-networking/translation/istio/extensions/istio_networking_extender.go:22]: swithced from a per-resource patch model to a whole-snapshot patch model

File: api/networking/extensions/v1alpha1/networking_extensions.proto [hunk 2]
```diff
@@ -21,92 +21,85 @@ import "networking/v1alpha3/workload_entry.proto";
...
-    // generated for a given SMH Workload
-    rpc GetWorkloadPatches (WorkloadPatchRequest) returns (PatchList) {
-    }
-
-    // GetMeshPatches retrieves desired patches to the set of output Mesh config resources
-    // generated for a given SMH Mesh
-    rpc GetMeshPatches (MeshPatchRequest) returns (PatchList) {
+    // GetExtensionPatches fetches a set of patches to the output configuration from the Extensions server.
+    // The current discovery snapshot and translated outputs are provided in the ExtensionPatchRequest
+    rpc GetExtensionPatches (ExtensionPatchRequest) returns (ExtensionPatchResponse) {
```
Comment (ilackarms) [inline api/networking/extensions/v1alpha1/networking_extensions.proto:26]: here we are switching from a per-resource extension model to a batch extension per-resource. this is to ensure the server always has a full snapshot of resources from SMH.

File: api/networking/extensions/v1alpha1/networking_extensions.proto [hunk 3]
```diff
@@ -21,92 +21,85 @@ import "networking/v1alpha3/workload_entry.proto";
...
-    repeated GeneratedResource generated_resources = 2;
+// the parameters provided to the Extensions server when requesting patches
+message ExtensionPatchRequest {
+    // the set of discovery objects provided as inputs for the SMH translation
+    DiscoverySnapshot inputs = 1;
+
+    // the base set of output objects translated by SMH.
+    // these may have been operated upon by a previous Extension server if multiple servers
+    // have been configured.
+    repeated GeneratedObject outputs = 2;
```
Comment (ilackarms) [inline api/networking/extensions/v1alpha1/networking_extensions.proto:45]: note: we will probably want to optimize this API in the future to reduce the grpc I/O. currently the server is dumb and expects an entire snapshot (both the discovery inputs and pre-patch translated outputs) in the patch request.

in the future we can make SMH client smarter to only send the updated objects, and have the server maintain a local cache

File: pkg/mesh-networking/extensions/conversions.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
+package extensions
+
+import (
+	discoveryv1alpha2 "github.com/solo-io/service-mesh-hub/pkg/api/discovery.smh.solo.io/v1alpha2"
+	"github.com/solo-io/service-mesh-hub/pkg/api/networking.smh.solo.io/extensions/v1alpha1"
+	"github.com/solo-io/service-mesh-hub/pkg/api/networking.smh.solo.io/input"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
```
Comment (ilackarms) [inline pkg/mesh-networking/extensions/conversions.go:9]: this file adds utility funcs shared by both client and server (converting from grpc to kube types for the wire)

File: pkg/mesh-networking/translation/istio/istio_networking_translator.go [hunk 5]
```diff
@@ -63,27 +61,7 @@ func (t *istioTranslator) Translate(
...
-
-		// TODO(ilackarms): add translation for workloads when a feature requires us to do so
-		//workloadTranslator.Translate(in, workload, workloadOutputs, reporter)
-
-		if err := t.extensions.PatchWorkloadOutputs(ctx, workload, workloadOutputs); err != nil {
-			contextutils.LoggerFrom(ctx).Errorf("failed to apply extension patches for traffic target %v", sets.Key(workload))
-		}
-
-		istioOutputs.Merge(workloadOutputs)
+		trafficTargetTranslator.Translate(in, trafficTarget, istioOutputs, reporter)
```
Comment (ilackarms) [inline pkg/mesh-networking/translation/istio/istio_networking_translator.go:64]: patching now happens once, at the end of translation

File: test/e2e/env.go [hunk 6]
```diff
@@ -219,11 +222,17 @@ func StartEnv(ctx context.Context) Env {
...
 	}
 
-	cmd := exec.CommandContext(ctx, "./ci/setup-kind.sh", strconv.Itoa(GinkgoParallelNode()))
+	// change to repo root dir
+	err := os.Chdir(filepath.Join(util.MustGetThisDir(), "..", ".."))
+	Expect(err).NotTo(HaveOccurred())
+
+	// get absolute path to setup script so this function can be called from any working directory)
+	setupScript := filepath.Join(util.MustGetThisDir(), "..", "..", "ci", "setup-kind.sh")
+	cmd := exec.CommandContext(ctx, setupScript, strconv.Itoa(GinkgoParallelNode()))
```
Comment (ilackarms) [inline test/e2e/env.go:231]: this change (chdir to smh repo) allows StartEnv to be called from e2e tests in the smh-e repo

File: codegen/groups/groups.go [hunk 7]
```diff
@@ -10,53 +10,58 @@ import (
...
+		{Kind: "IssuedCertificate"},
+		{Kind: "CertificateRequest"},
+		{Kind: "PodBounceDirective", NoStatus: true},
 	}),
 }
 
-type resourceToGenerate struct {
-	kind     string
-	noStatus bool // don't put a status on this resource
+type ResourceToGenerate struct {
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply codegen/groups/groups.go:41]: yes, in smh-enterprise

**Review by harveyxia — APPROVED** [review]

**Pull request merged**